src/bytes.ts

```diff
@@ -0,0 +1,265 @@
+import { assert } from './assert';
+import { add0x, assertIsHexString, remove0x } from './hex';
+
+export type Bytes = bigint | number | string | Uint8Array;
+
+/**
+ * Memoized function that returns an array to be used as a lookup table for
+ * converting bytes to hexadecimal values.
+ *
+ * The array is created lazily and then cached for future use. The benefit of
+ * this approach is that the performance of converting bytes to hex is much
+ * better than if we were to call `toString(16)` on each byte.
+ *
+ * The downside is that the array is created once and then never garbage
+ * collected. This is not a problem in practice because the array is only 256
+ * elements long.
+ *
+ * @returns A function that returns the lookup table.
+ */
+function getPrecomputedHexValuesBuilder(): () => string[] {
+  // To avoid issues with tree shaking, we need to use a function to return the
+  // array. This is because the array is only used in the `bytesToHex` function
+  // and if we were to use a global variable, the array might be removed by the
+  // tree shaker.
+  const lookupTable: string[] = [];
+
+  return () => {
+    if (lookupTable.length === 0) {
+      for (let i = 0; i < 256; i++) {
+        lookupTable.push(i.toString(16).padStart(2, '0'));
+      }
+    }
+
+    return lookupTable;
+  };
+}
+
+/**
+ * Function implementation of the {@link getPrecomputedHexValuesBuilder}
+ * function.
+ */
+const getPrecomputedHexValues = getPrecomputedHexValuesBuilder();
+
+/**
+ * Check if a value is a `Uint8Array`.
+ *
+ * @param value - The value to check.
+ * @returns Whether the value is a `Uint8Array`.
+ */
+export function isBytes(value: unknown): value is Uint8Array {
+  return value instanceof Uint8Array;
+}
+
+/**
+ * Assert that a value is a `Uint8Array`.
+ *
+ * @param value - The value to check.
+ * @throws If the value is not a `Uint8Array`.
+ */
+export function assertIsBytes(value: unknown): asserts value is Uint8Array {
+  assert(isBytes(value), 'Value must be a Uint8Array.');
+}
+
+/**
+ * Convert a `Uint8Array` to a hexadecimal string.
+ *
+ * @param bytes - The bytes to convert to a hexadecimal string.
+ * @returns The hexadecimal string.
+ */
+export function bytesToHex(bytes: Uint8Array): string {
+  assertIsBytes(bytes);
+
+  const lookupTable = getPrecomputedHexValues();
+  const hex = Array.prototype.map.call(bytes, (n) => lookupTable[n]).join('');
```

Comment:
`map` requires a function call for each element.

```typescript
const hex = new Array(bytes.length);
// probably could also be speed up with for (let index) instead of for-of
for (const n of bytes) {
  hex[i] = lookupTable[n];
}
return add0x(hex);
```